clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/Types.hs

```diff
@@ -0,0 +1,161 @@
+  AutoNegConfig
+  { cLinkSpeed :: LinkSpeed
+  , cDuplexMode :: DuplexMode
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Converts the AutoNegConfig Records to the an_adv_config_vector format.
+toAutoNegConfigVector :: AutoNegConfig -> AutoNegConfigVector
+toAutoNegConfigVector AutoNegConfig{..} = bv4
+  where
+    (linkSpeed0, linkSpeed1) = case cLinkSpeed of
+      Speed10 -> (0, 0)
+      Speed100 -> (0, 1)
+      Speed1000 -> (1, 0)
+
+    bv0 = 0x4001 -- How the bridge is configured in MAC mode.
+
+    -- Apply the configuration.
+    bv1 = replaceBit (12 :: Int) (bitCoerce $ cDuplexMode == FullDuplex) bv0
+    bv2 = replaceBit (10 :: Int) linkSpeed0 bv1
+    bv3 = replaceBit (11 :: Int) linkSpeed1 bv2
+
+    -- Set link status
+    bv4 = setBit bv3 (15:: Int)
+
+
+-- | 10 Mbit and half duplex are the default values since they are the most likely
+-- to be supported by the link partner.
+instance Default AutoNegConfig where
+  def = AutoNegConfig {..}
+   where
+    cLinkSpeed = Speed10
+    cDuplexMode = HalfDuplex
+
+type ConfigVector = BitVector 5
+-- | A record representation of the configuration_vector.
+data Config = Config
+  { cUniDirectional :: Bool
+  , cLoopback :: Bool
+  , cPowerDown :: Bool
+  , cIsolateGmii :: Bool
+  , cAutoNegEnable :: Bool
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Transform the Config record to a configuration_vector.
+toConfigVector :: Config -> ConfigVector
+toConfigVector Config{..} = pack
+  ( cAutoNegEnable
+  , cIsolateGmii
+  , cLoopback
+  , cPowerDown
+  , cUniDirectional
+  )
+
+instance Default Config where
+  def = Config {..}
+   where
+    cUniDirectional = False
+    cLoopback = False
+    cPowerDown = False
+    cIsolateGmii = False
+    cAutoNegEnable = True
+
+type StatusVector = BitVector 16
+-- | A record representation of the status_vector.
+data Status = Status
+  { sLinkStatus :: Bool
+  , sLinkSynchronized :: Bool
+  , sRudiC :: Bool
+  , sRudiI :: Bool
+  , sRudiInvalid :: Bool
+  , sReceivedDisparityError :: Bool
+  , sReceivedInvalidCode :: Bool
+  , sPhyLinkStatus :: Bool
+  , sRemoteFault :: Maybe (BitVector 2)
+  , sLinkSpeed :: LinkSpeed
+  , sDuplexMode :: DuplexMode
+  , sPause :: Pause
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Transform the status_vector to a Status record.
+fromStatusVector :: StatusVector -> Status
+fromStatusVector v = Status{..}
+ where
+  -- There is no BitPack instance for 13-tuples without -flarge-tuples, so we break it up.
+  ( ( sLinkStatus
+    , sLinkSynchronized
+    , sRudiC
+    , sRudiI
+    , sRudiInvalid
+    , sReceivedDisparityError
+    )
+    , sReceivedInvalidCode
+    , sPhyLinkStatus
+    , remoteFaultCode
+    , linkSpeedBools
+    , duplexModeBool
+    , remoteFaultValid
+    , pauseBools) = unpack v
+  sRemoteFault = if remoteFaultValid then Just remoteFaultCode else Nothing
+  sLinkSpeed = case linkSpeedBools of
+    (True, False) -> Speed1000
+    (False, True) -> Speed100
+    (False, False) -> Speed10
```

Comment:
It would be nice if this could be done with `unpack @LinkSpeed`, but failing that, surely it could be done with `linkSpeedBV :: BitVector 2` rather than booleans?
```suggestion
  sLinkSpeed = case linkSpeedBV of
    0 -> Speed10
    1 -> Speed100
    2 -> Speed1000
```